File: 0301-0400/0300.Longest_Increasing_Subsequence.py
#1245ms





class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        n = len(nums)
        dp = [1] * n

        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)

        return max(dp)








#0ms








class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        
        i = 0
        l = len(nums)
        ans = []
        while i<l:
            if len(ans) == 0 or nums[i]>ans[-1]:
                ans.append(nums[i])
            else:
                idx = bisect_left(ans, nums[i])
                ans[idx]= nums[i]
            i+=1
        return len(ans)








#43ms









class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        best_chain = []
        for num in nums:
            if not best_chain or num > best_chain[-1]:
                best_chain.append(num)                
            else:
                left = 0
                right = len(best_chain)-1
                while left <= right:
                    middle = (left+right) // 2
                    if best_chain[middle] < num:
                        left += 1
                    else:
                        right -= 1
                best_chain[left] = num
        return len(best_chain)
        

File: 0301-0400/0300.Longest_Increasing_Subsequence.java
//51ms






public class Solution {
    public int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int n = nums.length;
        int[] dp = new int[n];
        Arrays.fill(dp, 1);

        for (int i = 1; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
        }

        int maxLength = Arrays.stream(dp).max().orElse(0);
        return maxLength;
    }
}





//10ms






class Solution{
    public int lengthOfLIS(int[] nums){
        List<Integer> list = new ArrayList<>();
        list.add(nums[0]);

        for(int i = 1; i < nums.length; i++){

            if(nums[i] > list.get(list.size() - 1)){
                list.add(nums[i]);
            }
            else{
                int left = 0;
                int right = list.size() - 1;

                while(left <= right){
                    int mid = left + (right - left) / 2;

                    if(list.get(mid) >= nums[i]){
                        right = mid - 1;
                    }
                    else{
                        left = mid + 1;
                    }
                }

                list.set(left, nums[i]);
            }
        }

        return list.size();
    }
}
